src/ripple/app/paths/impl/BookStep.cpp

```diff
@@ -546,11 +552,14 @@ BookStep<TIn, TOut>::check(StrandContext const& ctx) const
         JLOG (j_.debug()) << "Book: currency is inconsistent with issuer." << *this;
         return temBAD_PATH;
     }
-    if (!ctx.seenBooks.insert (book_).second)
+
+    if (!ctx.seenBookOuts.insert (book_.out).second ||
+        ctx.seenDirectIssues[0].count (book_.out))
     {
         JLOG (j_.debug()) << "BookStep: loop detected: " << *this;
```

Comment:
Personally, I think it would be worth adding a paragraph comment about why seeing the same `directIssue` twice constitutes a loop.  The write to the log is sufficient for the log, but it is not sufficient for understanding the problem.  Remember the ah ha we both had while talking to David about this problem.  Give that ah ha to the next person reading this code.